Set SHELL env var in launcher if unset to avoid issues in Code's shell detection

amisevsk opened the pull request:
### What does this PR do?
If the SHELL environment variable is unset when running the Code launcher, set it:
* If we detect that bash is installed, to `/bin/bash`
* Otherwise, fallback to setting it to `/bin/sh`

This is necessary due to shell-detection logic within Code itself, which will fallback to parsing /etc/passwd when SHELL is not set (see [[1]](https://github.com/che-incubator/che-code/blob/a2bf97a1322166c72127a5b17a5eced40e803fbf/code/src/vs/base/node/shell.ts#L42-L48)). When running on e.g. OpenShift with a normal container, cri-o will add an /etc/passwd entry for the current user with `/sbin/nologin`, which results the terminal failing to launch.

### What issues does this PR fix?
Closes https://github.com/eclipse/che/issues/22524

### How to test this PR?
Changes from this PR are built into `quay.io/amisevsk/che-code:dev` via the following Dockerfile:
```dockerfile
FROM registry.access.redhat.com/ubi8/nodejs-18:1-71 as launcher-builder

USER root

COPY . /checode-launcher
WORKDIR /checode-launcher

RUN npm install -g yarn@1.22.17

RUN yarn \
    && mkdir /test-launcher \
    && cp -r out/src/*.js /test-launcher \
    && chgrp -R 0 /test-launcher && chmod -R g+rwX /test-launcher

FROM quay.io/che-incubator/che-code:next
COPY --from=launcher-builder --chown=0:0 /test-launcher /checode-linux-libc/launcher
```

To test the changes on an existing workspace, edit the DevWorkspace yaml on the cluster, updating the `.spec.contributions` field:
```diff
  spec:
    contributions:
      - name: editor
        kubernetes:
          name: che-code-java-maven
+       components:
+         - container:
+             image: 'quay.io/amisevsk/che-code:dev'
+           name: che-code-injector
```
(This will use quay.io/amisevsk/che-code:dev in place of the default injector image)

We should verify three cases before merging:
1. Samples that do not set the `SHELL` environment variable have https://github.com/eclipse/che/issues/22524 resolved.
     * To make this easier (since links to raw devfiles in registry.devfile.io do not seem to work in Che), you should be able to use the raw link on this gist and paste it into the dashboard: https://gist.github.com/amisevsk/c4ffd5185d81420be2d8e73cc1db9d45 (taken from https://registry.devfile.io/viewer/devfiles/community/java-maven), then update the DevWorkspace as above
2. It's still possible to override the shell by setting the SHELL env var in the devfile or container
3. Default Che samples are unimpacted

amisevsk committed:
Detect shell in launcher and set env var if unset

Before launching Code, if the SHELL environment variable is unset, set
it to something before launching:

* If SHELL is unset and /bin/bash is installed, use that
* Otherwise, fall back to /bin/sh as a 'safe' option

This is necessary to avoid internal shell-detection logic in Code, which
reads /etc/passwd when SHELL is not set. This is an issue when running
in e.g. OpenShift, where cri-o will add a user entry to /etc/passwd
containing /sbin/nologin as the login shell.

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

amisevsk committed:
Add test cases to cover setting SHELL env var in launcher

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

amisevsk committed:
Fixup typo in workspace file creation logs

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

l0rd commented on launcher/src/vscode-launcher.ts:
```diff
@@ -66,6 +66,17 @@ export class VSCodeLauncher {
       env.NODE_EXTRA_CA_CERTS = NODE_EXTRA_CERTIFICATE;
     }
 
+    if (!env.SHELL) {
```
What about checking if the user shell (in `/etc/passwd`) is /sbin/nologin or not? If cri-o hasn't patched `/etc/passwd` (i.e. containerd is the container runtime) there may be a valid shell for current user anyway.

Beyond that I have tested the PR using image `quay.io/che-incubator-pull-requests/che-code:pr-284-amd64` and the [quarkus devfile from registry.devfile.io](https://registry.devfile.io/devfiles/java-quarkus/1.3.0) and the problem got solved 🎉 . I had an unrelated problem though and [I have filed a separate issue](https://github.com/eclipse/che/issues/22601).

amisevsk replied:
This is probably a good idea -- I didn't initially do it since I was trying to keep it simple (since IIRC the default UDI image sets SHELL) but better safe than sorry :)

amisevsk replied:
Updated the PR now. If /etc/passwd contains anything other than `/sbin/nologin`, we no longer override it by setting SHELL if it's not already set.

amisevsk committed:
Check /etc/passwd in launcher and only set SHELL if needed

When launching Code, add an additional check to read /etc/passwd and
check if the user has a (non-/sbin/nologin) shell defined there. If
/etc/passwd contains any other shell (e.g. /bin/zsh, /bin/bash, etc.),
then let Code read that shell from /etc/passwd instead of overriding it
by setting the SHELL environment variable.

Signed-off-by: Angel Misevski <amisevsk@redhat.com>

amisevsk commented:
@vitaliy-guliy @l0rd Updated this PR after Mario's suggestion, please take a look before I merge.

devstudio-release commented:
Build [3.10](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [code_3.x/959](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/code_3.x/959/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/code_3.x/959/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/code_3.x/959/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/code_3.x/959/git)

devstudio-release commented:
Build [3.10](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [sync-to-downstream_3.x/4970](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/4970/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/4970/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/4970/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/sync-to-downstream_3.x/4970/git)

devstudio-release commented:
Build [3.10](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [get-sources-rhpkg-container-build_3.x/4794](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/): [Console](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/console), [Changes](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/changes), [Git Data](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/git)

devstudio-release commented:
Build [3.10](https://github.com/redhat-developer/devspaces-images/tree/devspaces-3-rhel-8/) :: [get-sources-rhpkg-container-build_3.x/4794](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/): [![FAILURE](https://img.shields.io/badge/Build-FAILURE-red?style=plastic&logo=redhat)](https://main-jenkins-csb-crwqe.apps.ocp-c1.prod.psi.redhat.com/job/DS_CI/job/get-sources-rhpkg-container-build_3.x/4794/)<br /><blockquote>code : 3.x ::  <br /> ; copied to quay</blockquote>
